### ISSUE TITLE
Confusing 2D collision behavior on CanvasLayer with moving camera

### ISSUE THREAD

**Issue by ekopari** [godotengine/godot#86829]
### Tested versions

Reproducible in 4.2.1, 3.5.3 and 4.3 dev 1

### System information

Godot v4.2.1.stable - Arch Linux #1 SMP PREEMPT_DYNAMIC Thu Jan  4 09:50:36 EET 2024 - Wayland - Vulkan (Forward+) - integrated Intel(R) Graphics (RPL-P) () - 13th Gen Intel(R) Core(TM) i7-13700H (20 Threads)

### Issue description

Collisions occur between objects in different `CanvasLayers`, although the objects do not appear to touch, when the camera is following one of the objects. It seems the engine is computing the collisions based on the object's absolute/initial coordinates, not taking into account `CanvasLayer` effects and (visual) transforms.

Seems related to [bug 58514](https://github.com/godotengine/godot/issues/58514)

https://github.com/godotengine/godot/assets/44967740/7d9e8b4e-4db4-4866-b3c1-924c48a4a988

### Steps to reproduce

1. Create a new 2D scene.
2. As subnodes of the root node:
3. Create a `CanvasLayer`. Add a `StaticBody2D` as a subnode of the canvas layer, with a suitable collision shape.
4. Add a `RigidBody2D` with any collision shape and position it so that it lies "above" (lower Y coordinate) the `StaticBody2D` in the `CanvasLayer`. Add a `Camera2D` as a subnode of the rigid body and make sure it is enabled.
5. Create another `StaticBody2D` with a suitable collision shape and place it "below"(higher Y coordinate) the first `StaticBody2D`.
6. Activate visible collision shapes
7. Run the project and see the the rigid body apparently stop mid-air.

### Minimal reproduction project (MRP)

[UILayerCollisionBugDemo.zip](https://github.com/godotengine/godot/files/13842095/UILayerCollisionBugDemo.zip)

**Comment by Sauermann** [comment]
Currently only collisions of objects within a single canvas layer are supported. See the description of [CollisionObject2D](https://docs.godotengine.org/en/stable/classes/class_collisionobject2d.html).

One way to prevent these wrong collisions in different canvas layers is to put the `CollisionObjet2D`-nodes into different `collision_layer`s.

**Comment by ekopari** [comment]
Of course there are workarounds, but I think the main problem is that it can be very confusing.
I discovered this problem when my player node was colliding with something - the `body_entered` signal was getting called, but activating visible collision shapes to quickly see what it was colliding with didn't reveal anything useful.

**Comment by Sauermann** [comment]
I agree, that this behaviour needs to be improved.
The last attempt to fix this got reverted (#59881), but while trying this, it became apparent, that many locations in the code base need to be adjusted and this needs to be considered a complex and risky change.

**Comment by ballerburg9005** [comment]
https://github.com/godotengine/godot/assets/50874674/25bb6cec-ca6f-449c-9139-3ae92a90ebff

So what is the best workaround here?

The best idea I came up with in 30 minutes was to make a clone of the object in both layers, with layer masks such that they only apply to the layer they are on. Then collisions in the lower layers are blocked by custom code, if there had been successful collisions on the upper layers.

**Comment by ekopari** [comment]
@ballerburg9005 I am not sure I understand what you are trying to achieve, but is seems you are trying to implement some form of drag-and-drop mechanism between a fixed UI and the game world. The base problem may be the same, but in my case I could solve my problem by completely disabling collisions on my fixed UI nodes. I don't think this would work in your case, at least not by itself. The only idea I have that could help in your situation would be to create a global drag-and-drop controller node ([singleton/autoload](https://docs.godotengine.org/en/stable/tutorials/scripting/singletons_autoload.html)) that would record the currently dragged item and that could be queried by any interested node when it wants to read the currently dragged item - for example on mouse entered and mouse release.

**Comment by ballerburg9005** [comment]
Yes, like you can see very simply being able to drag&drop items from the HUD to the game world, it requires collision to work properly.

If I understand you correctly, you propose that I change my dragndrop code, so that it is not based on actually dragging a world item that collides, but based on mouse events only. I don't really know, I think that produces lots of other issues down the road, since there might be physics interactions while dragging or the collision shape might vary.

I went for the clone solution. So once you drag from whatever place, it spawns not one but two game-world items into drag-space, but with individual collision layer/mask seperated between UI and game world, each respectively parented to either game-world or HUD node that run on differnt canvas layers... so the broken inter-canvaslayer collisions are not possible anymore (but you still get all the collisions, just seperated between two items). It made the code pretty ugly to look at, but it is probably the most sane way to deal with this bug/limitation. And also when this is fixed you can just clean the mess up and get rid of this double-spawn code fairly easily.

**Comment by Nikola526** [comment]
I have a 2d top-down game where I was counting on this behavior to implement different floors of a house. While the lighting is scoped to its own canvaslayer, the physics is not and my character is colliding with the physics of the layer underneath. Toggling layer bits when the character moves a level sounds problematic and bug-prone.

**Comment by ballerburg9005** [comment]
@Nikola526 This bug (if I remember correctly) is about the collision physics of the canvas layer being fixed to 0,0 coordinates in the game world. Thus when the camera moves, what the UI canvaslayer actually draws mismatches by whatever camera movement offset between both layers.

Vice versa you can imagine the UI layer's physics to be fixed in place in game-world coordinates.

<img width="1024" height="1402" alt="Image" src="https://github.com/user-attachments/assets/cc64a836-32a5-409a-b852-889f194d038f" />
